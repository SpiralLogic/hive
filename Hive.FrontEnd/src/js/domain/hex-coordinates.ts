export type HexCoordinates = {
    readonly q: number;
    readonly r: number;
};
