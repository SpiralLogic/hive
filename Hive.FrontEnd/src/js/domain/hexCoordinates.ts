export interface HexCoordinates {
  q: number;
  r: number;
}