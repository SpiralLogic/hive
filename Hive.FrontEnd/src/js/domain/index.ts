export * from './cell';
export * from './game-state';
export * from './move';
export * from './tile';
export * from './player';
export * from './cell';
export * from './hex-coordinates';
