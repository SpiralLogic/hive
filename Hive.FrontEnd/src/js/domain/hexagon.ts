import { Tile } from './tile';
import { HexCoordinates } from './hex-coordinates';

export interface Hexagon {
    readonly coordinates: HexCoordinates;
    readonly tiles: Tile[];
}
