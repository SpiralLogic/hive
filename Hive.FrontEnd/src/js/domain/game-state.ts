import { Hexagon } from './hexagon';
import { Player } from './player';

export interface GameState {
    hexagons: Hexagon[];
    players: Player[];
}
