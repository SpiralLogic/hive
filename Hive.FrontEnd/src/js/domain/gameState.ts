import { Hexagon } from './hexagon';
import { Player, PlayerId } from './player';

export interface GameState {
  hexagons: Hexagon[];
  players: Player[];
}