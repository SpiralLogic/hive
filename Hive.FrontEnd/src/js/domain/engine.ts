import { GameState } from './game-state';
import { Move } from './move';

export interface HexEngine {
    newGame(): Promise<GameState>;
    move(move: Move): Promise<GameState>;
}
