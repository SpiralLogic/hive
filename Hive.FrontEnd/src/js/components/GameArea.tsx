import { h } from 'preact';
import { useState } from 'preact/compat';
import { useEffect } from 'preact/hooks';
import { GameState } from '../domain';
import { CellDropEvent, useCellDropEmitter } from '../emitters/cell-drop-emitter';
import Engine from '../game-engine';
import { handleDragOver } from '../handlers';
import Hextille from './Hextille';
import PlayerList from './PlayerList';

const cellDropEmitter = useCellDropEmitter();

const GameArea = () => {
    const [gameState, setGameState] = useState<GameState | undefined>(undefined);

    const attributes = {
        ondragover: handleDragOver,
        className: 'hive',
        style: { '--hex-size': '50px' },
    };

    useEffect(() => {
        const fetch = (async () => {
            setGameState(await Engine.newGame());
        });
        fetch();
    }, []);

    useEffect(() => {
        const cellDropListener = async (e: CellDropEvent) => setGameState(await Engine.moveTile(e));

        cellDropEmitter.add(cellDropListener);
        return () => cellDropEmitter.remove(cellDropListener);
    }, [gameState]);

    if (!gameState) {
        return <h1>loading !</h1>;
    }

    return (
        <div {...attributes}>
            <PlayerList players={gameState.players}/>
            <Hextille cells={gameState.cells}/>
        </div>
    );
};

GameArea.displayName = 'GameArea';
export default GameArea;
