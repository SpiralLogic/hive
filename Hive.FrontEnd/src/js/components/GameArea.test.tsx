import { GameArea } from './GameArea';
import { act, create } from 'react-test-renderer';
import * as React from 'react';
import { Hexagon } from '../domain';

jest.mock('../domain/engine');
const hexagon: Hexagon = { coordinates: { q: 1, r: 1 }, tiles: [] };
const gameState = Promise.resolve({ hexagons: [hexagon], players: [] });

test('Game area shows loading', () => {
    const state = jest.fn().mockImplementation(() => gameState);
    const move = jest.fn().mockImplementation(() => gameState);

    const engine = { initialState: state, moveTile: move };
    const component = create(<GameArea engine={engine}/>);
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});

test('Game shows board after fetch', async () => {
    const state = jest.fn().mockImplementation(() => gameState);
    const move = jest.fn().mockImplementation(() => gameState);

    const engine = { initialState: state, moveTile: move };
    const component = create(<GameArea engine={engine}/>);

    await act(async () => { await state;});
    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
});
