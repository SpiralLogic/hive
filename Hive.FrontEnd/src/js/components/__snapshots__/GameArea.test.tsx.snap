// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game area shows loading 1`] = `
<h1>
  loading
</h1>
`;

exports[`Game shows board after fetch 1`] = `
<div
  className="hive"
  style={
    Object {
      "--hex-size": "50px",
    }
  }
>
  <div
    className="players"
  />
  <div
    className="hextille"
    style={
      Object {
        "--hex-offset": -1,
      }
    }
  >
    <div
      className="hex-row"
    >
      <div
        className="hidden"
      />
      <div
        className="hex cell"
      />
    </div>
  </div>
</div>
`;
